src/protocol/mod.rs

```diff
@@ -149,6 +149,12 @@ enum WriteState {
     WriteData(VecWriter),
 }
 
+enum NextWriteState {
+    Same,
+    None,
```

Comment:
Can we call this `Nothing` or `Done`? My brain interprets `None` as being a variant of `Option`.